File: CIS194/homework1/solutions.scala
// Exercise 1
def toDigits(x: Long): List[Long] = {
  def toDigitList(lst: List[Char]): List[Long] = 
    lst match {
      case Nil => Nil
      case x :: xs => x.asDigit :: toDigitList(xs)
    }
  x match {
    case n if n <= 0 => Nil
    case n => toDigitList(x.toString.toList)
  }
}

def toDigitsRev(x: Long): List[Long] = 
  toDigits(x).reverse
// End Exercise 1

// Exercise 2
def doubleEveryOther(Longs: List[Long]): List[Long] =
  Stream.from(1).
    zip(Longs.reverse).
    map(x => if (x._1 % 2 == 0) x._2 * 2 else x._2).
    toList.
    reverse
// End Exercise 2

// Exercise 3
def sumDigits(Longs: List[Long]): Long = 
  Longs.flatMap(toDigits).sum
// End Exercise 3


// Exercise 4
def validate(x: Long): Boolean = {
  val sumAll: Long => Long = 
    (toDigits _) andThen
    doubleEveryOther andThen
    sumDigits

  x match {
    case n if n <= 0 => false
    case n => if (sumAll(n) % 10 == 0) true else false
  }
}

// End Exercise 4


// Exercise 5
type Peg = String
type Move = (Peg, Peg)
def hanoi(n: Int, a: Peg, b: Peg, c: Peg): List[Move] = 
  n match {
    case 0 => Nil
    case n => hanoi(n - 1, a, c, b) ++ 
              List((a, b)) ++ 
              hanoi(n - 1, c, b, a)
  }

// End Exercise 5


// Exercise 6 (Optional)
def hanoi4(n: Int, a: Peg, b: Peg, c: Peg, d: Peg): List[Move] = 
  n match {
    case 0 => Nil
    case n => hanoi4(n / 2, a, c, b, d) ++ 
              hanoi(n - (n/2), a, b, d) ++
              hanoi4(n / 2, c, b, a, d)
  }
// End Exercise 6

File: CIS194/homework1/solutions.hs
-- Exercise 1
--
toDigits  :: Integer -> [Integer]
toDigits x
    | x == 0    = []
    | x < 0     = []
    | otherwise = map read $ map (\x -> [x]) (show x)

toDigitsRev :: Integer -> [Integer]
toDigitsRev = reverse . toDigits

-- End Exercise 1


-- Exercise 2
--
doubleEveryOther :: [Integer] -> [Integer]
doubleEveryOther = reverse . map everyOther . zip [1..] . reverse 
                where everyOther x = 
                        if mod (fst x) 2 == 0 
                        then (snd x)*2 
                        else snd x

-- End Exercise 2


-- Exercise 3
--
sumDigits :: [Integer] -> Integer
sumDigits []     = 0
sumDigits [x]    = if x < 10 then x else sumDigits (toDigits x)
sumDigits (x:xs) = (sumDigits [x]) + (sumDigits xs)

-- End Exercise 3


-- Exercise 4
--
validate :: Integer -> Bool
validate x
    | x < 0 = False
    | x == 0 = False
    | otherwise = if (sumAll x) `mod` 10 == 0 then True else False
    where sumAll = sumDigits . doubleEveryOther . toDigits 

-- End Exercide 4


-- Exercise 5
type Peg = String
type Move = (Peg, Peg)
hanoi :: Integer -> Peg -> Peg -> Peg -> [Move]
hanoi 0 _ _ _ = []
hanoi n a b c = hanoi (n-1) a c b ++ [(a, b)] ++ hanoi (n-1) c b a

-- End Exercise 5


-- Exercise 6 (Optional)
--
hanoi4 :: Integer -> Peg -> Peg -> Peg -> Peg -> [Move]
hanoi4 0 _ _ _ _ = []
hanoi4 n a b c d = 
    hanoi4 m a c b d ++ hanoi (n - m) a b d ++ hanoi4 m c b a d
    where m = quot n 2

-- End Exercise 6

-- Extra (Generalized case of N Disks and K pegs.)
    -- Code modified from: http://stackoverflow.com/questions/3607161/towers-of-hanoi-with-k-pegs
hanoiK :: Int -> [a] -> [(a, a)]
hanoiK 0 _ = []
hanoiK 1 (a : b : rest) = [(a, b)]
hanoiK n (a:b:c:rest)
    | null rest = 
            hanoiK (n - 1) (a : c : b : rest) ++
            hanoiK 1 (a : b : rest) ++
            hanoiK (n - 1) (c : b : a : rest)
    | otherwise = 
            hanoiK (quot n 2) (a : c : b : rest) ++
            hanoiK (n - (quot n 2)) (a : b : rest) ++
            hanoiK (quot n 2) (c : b : a : rest)
